node/core/dispute-coordinator/src/initialized.rs

```diff
@@ -715,18 +676,21 @@ impl Initialized {
 		session: SessionIndex,
 		statements: Vec<(SignedDisputeStatement, ValidatorIndex)>,
 		now: Timestamp,
+		leaf_hash: Hash,
 	) -> FatalResult<ImportStatementsResult> {
 		gum::trace!(target: LOG_TARGET, ?statements, "In handle import statements");
-		if !self.rolling_session_window.contains(session) {
+		if self.session_is_ancient(session) {
 			// It is not valid to participate in an ancient dispute (spam?) or too new.
 			return Ok(ImportStatementsResult::InvalidImport)
 		}
 
-		let env = match CandidateEnvironment::new(
-			&self.keystore,
-			&self.rolling_session_window,
-			session,
-		) {
+		let session_info = &self
+			.runtime_info
+			.get_session_info_by_index(ctx.sender(), leaf_hash, session)
```

Comment:
By above reasoning, better to use the relay_parent of the candidate here. This way cache pre-filling will use a different relay parent than we are using here, maximizing our chanced of at least one succeeding.

Reply:
The parameter name is misleading. `handle_import_statements` is called in multiple contexts:
- `process_onchain_votes` (with leaf hash)
- `handle_incoming` (with relay parent)
- `issue_local_statement` (with relay parent)

I'll fix the parameter name and we should be good.

Reply:
Maybe we should get rid of this redundancy?

Reply:
> assumed to exist already in the db.

I can't find this in the code. Can you share a link?

Reply:
[here](https://github.com/paritytech/polkadot/blob/97bd688932499e309c085f67298ad3deb98b7420/node/core/dispute-coordinator/src/initialized.rs#L760) either the candidate receipt is provided by the caller or it must be already present in the db, otherwise the import fails. Hence we always have access to the relay parent and passing in the leaf explicitly is redundant.

Reply:
As far as I can see the `CandidateReceipt` is not persisted in the db: 
https://github.com/paritytech/polkadot/blob/c2775a6d1e2c2931c14e8c049673bda2a731b695/node/core/dispute-coordinator/src/backend.rs#L66

And if `MaybeCandidateReceipt` contains `AssumeBackingVotePresent` - we can't recover the receipt. 

What am I missing here?

Reply:
It is, in [`CandidateVotes`](https://github.com/paritytech/polkadot/blob/27ddd270c3979fd2a5e2ac6fb3c9d474962c9260/node/primitives/src/disputes/mod.rs#L53)